### ISSUE TITLE
WebExtensions.WebExistsFullUrl returns true for non-existent sites from different managed path in Sharepoint 2013 on-premise

### ISSUE THREAD

**Issue by sadomovalex** [pnp/PnP-Sites-Core#1634]
#### Category
[ x] Bug
[ ] Enhancement

#### Environment
[ ] Office 365 / SharePoint Online
[ ] SharePoint 2016
[ x] SharePoint 2013

If SharePoint on-premises, what's exact CU version: 
- May 2015 CU (15.0.4719.1002)

#### Expected or Desired Behavior
It looks like WebExtensions.WebExistsFullUrl method work properly only within single site collection i.e. when client context (which is extended with this extension method) and url to check belong to the same managed path.

Example:
context is created from the root site http://example.com. This site has http://example.com/test sub site. In this case WebExistsFullUrl returns true for http://example.com/test url and false for some non-existent sub site's url like http://example.com/test123. I.e. behavior is correct.

But if context and url to check belong to different managed path then it always returns true.

Example:
context is created from the root site http://example.com and we call WebExistsFullUrl for some url which may belong to other site collection (e.g. which uses different managed path like http://example.com/teams/some-not-real-url. Suppose that at the moment of call we don't know whether this collection exists or not - we want to determine it by WebExistsFullUrl call). In this case WebExistsFullUrl returns true even if site collection with specified url doesn't exists.

I analyzed the code and found the following. Here is current version of WebExistsFullUrl  extension method:
![](https://drive.google.com/uc?export=download&id=1ki0N3xO3y3VcOj7AAbTS4j85Kv5vhKbz)

When we call this method with context and url which belong to the same managed path it throws exception like expected. But when it is called with context and url which belong to different managed paths exception is not thrown. Instead context is created for the root site http://example.com and method returns true and caller thinks that site exists. In order to fix this problem for different managed paths I applied the following fix in our local version:
![](https://drive.google.com/uc?export=download&id=1ttMnFDv4IkGCnNbC8KMEsCeO0mJvNDe9)

I.e. before to return true method also checks whether loaded url is the same as url to be checked. If they are different (which happens in scenario with different managed paths) it will return false like it should. Os course may be this method was not supposed to be used on different site collections/managed paths, but I think it should not return true at least for non-existent sites.

PS. Described behavior was found when root site collection (http://example.com) was host-named site collection, but most probably it will also work same way for regular site collections.
PPS. Similar behavior was for regular object model in past Sharepoint versions - when we passed non-existent url to SPWeb constructor it didn't throw exception. Instead we needed to check SPWeb.Exists property to ensure that site really exists.

**Comment by armchairdeity** [comment]
If the method really is only relative to the current context's managed path then shouldn't trying to verify a URL outside that managed context throw an error like IsOutsideManagedWebPathException or something?

It shouldn't return a result at all if it can't actually get to the URL that we're asking it to verify.

**Comment by armchairdeity** [comment]
We are having this exact same issue just now... I REALLY DO NOT want to have to clone the repo, edit, build, and maintain a copy for our build. That would suck.

So hey, PnP folks... can we please get a freakin fix? This bug was filed 9 months ago... holy crap, folks... a human woman's body can craft an entirely new human being in that length of time.

This is just one function. Surely SOMETHING can be done about this, and more quickly than it takes to forge a brand new baby?

**Comment by SchauDK** [comment]
@armchairdeity why don't you submit a PR to get this fixed and contribute to the community like the rest of us.

**Comment by armchairdeity** [comment]
@SchauDK - being as I don't know you, you don't know me, and I don't really want to get drawn into an argument today... the following is offered in blunt and straightforward speech, not intended to be snide, sarcastic or irritated. So... I will just answer your question as tho this is what you were actually asking:

I've been creating and contributing to OSS projects for 15 years, so it's not like I'm averse to helping out, in any way. I've seen some pretty good conversations about technique and approach on this project... I'd be happy to clone, fix and create a PR... except on this particular bug at this particular time.

But, on the other hand... if you're asking why don't I just step in and fix this particular bug at this particular time?

Well... probably because it's a complicated problem with several different solutions all of which have slightly different results and I don't know what the product team's vision for this function was in the first place, Because the OP proposed a solution in his post. Because none of the team members have seen fit to acknowledge, comment on, or close this bug in the 9 months it's been here.

Hell we don't even know if this is intended behavior or an actual bug or what the intended behavior really is... so how can I create a PR to fix it?

- Do we create a completely new context with the passed in URL?
- Or do we use a standard webrequest to check the status code?
- If we do that what if the credentials in the original context are not the same as the credentials for the site being queried? With invalid query credentials... what response will we get from Sharepoint? I know that if you do not have perms on the requested site you get a 302 from IIS... even if your credentials are correct. What if they aren't valid but the site exists? Should a 302 be considered a valid site? It had to be valid to return a redirect...
- What about a 403? Not authorized. You can't be not authorized to something that doesn't exist... so is that a positive response?
- What about a 500 vs a 404? A 404 means at least that the file, document, or image doesn't exist... but does the SITE? Shouldn't we be dropping the last element in script_path? Or not? OTOH, a 500 suggests that there's at least a live host at that connection... so... does the site exist? Well... maybe. We didn't get to talk to the server, it just tells us that the server exists.

Now if someone from the project team had seen fit to comment here and either acknowledge the proposed fix ,or stated some alternative approach, then sure... maybe I could do that. But I'm not going to take on such a potentially expansive (not to mention potentially breaking) task without some guidance. 

Basically what I'm saying is that the product team dropped the ball on this and they need, IMO, to at least give the TICKET some attention so that it CAN be fixed... because right now this issue is at the heart of my bug at work, thus impacting me, personally, directly, and I want some attention on it. Also because #mememe, and because I'm supposed to take my daughter fishing this afternoon and this bug that's been sitting around ignored for 9 months is screwing up daddy/daughter time.

So yeah... umm... that's why. ;)

**Comment by sadomovalex** [comment]
I checked this issue with the latest PnP version (which is 3.26.2010 at the moment of writing) and found that this problem is fixed now by MS. I.e. code of WebExtensions.WebExistsFullUrl method is still the same (see above) but now when we try to clone ClientContext in this line:
```
using (ClientContext testContext = context.Clone(webFullUrl))
{
...
}
```
and webFullUrl corresponds to non-existent site collection from another managed path (e.g. if context was created from the root site collection http://example.com and we pass webFullUrl = http://example.com/sites/som-non-existent-url) it throws System.Exception (System.Net.WebException) now: The remote server returned an error: (404) Not Found (in earlier CSOM version it created ClientContext for the root site in this case). And as result WebExtensions.WebExistsFullUrl returns false as expected. So issue can be closed.
